arangod/GeneralServer/H2CommTask.h

```diff
@@ -0,0 +1,129 @@
+#include <boost/lockfree/queue.hpp>
+#include <memory>
+
+namespace arangodb {
+class HttpRequest;
+class HttpResponse;
+
+namespace rest {
+
+template <SocketType T>
+class H2CommTask final : public GeneralCommTask<T> {
+ public:
+  H2CommTask(GeneralServer& server, ConnectionInfo, std::unique_ptr<AsioSocket<T>> so);
+  ~H2CommTask() noexcept;
+
+  void start() override;
+  /// @brief upgrade from  H1 connection, must not call start
+  void upgradeHttp1(std::unique_ptr<HttpRequest> req);
+
+ protected:
+  // set a read timeout in asyncReadSome
+  bool enableReadTimeout() const override { return true; }
+
+  bool readCallback(asio_ns::error_code ec) override;
+
+  void sendResponse(std::unique_ptr<GeneralResponse> response, RequestStatistics* stat) override;
+
+  std::unique_ptr<GeneralResponse> createResponse(rest::ResponseCode, uint64_t messageId) override;
+
+ private:
+  static int on_begin_headers(nghttp2_session* session,
+                              const nghttp2_frame* frame, void* user_data);
+  static int on_header(nghttp2_session* session, const nghttp2_frame* frame,
+                       const uint8_t* name, size_t namelen, const uint8_t* value,
+                       size_t valuelen, uint8_t flags, void* user_data);
+  static int on_frame_recv(nghttp2_session* session, const nghttp2_frame* frame,
+                           void* user_data);
+  static int on_data_chunk_recv(nghttp2_session* session, uint8_t flags, int32_t stream_id,
+                                const uint8_t* data, size_t len, void* user_data);
+  static int on_stream_close(nghttp2_session* session, int32_t stream_id,
+                             uint32_t error_code, void* user_data);
+  static int on_frame_send(nghttp2_session* session, const nghttp2_frame* frame,
+                           void* user_data);
+
+  static int on_frame_not_send(nghttp2_session* session, const nghttp2_frame* frame,
+                               int lib_error_code, void* user_data);
+
+ private:
+  // ongoing Http2 stream
+  struct Stream {
+    Stream(std::unique_ptr<HttpRequest> req) : request(std::move(req)) {}
+    
+    std::string origin;
+
+    std::unique_ptr<HttpRequest> request;
+    std::unique_ptr<HttpResponse> response;
+
+    size_t headerBuffSize = 0; // total header size
+    size_t responseOffset = 0; // current offset in response body
+  };
+
+  /// init h2 session
+  void initNgHttp2Session();
+
+  void processStream(Stream* strm);
+
+  /// should close connection
+  bool shouldStop() const;
+
+  // queue the response onto the session, call only on IO thread
+  void queueHttp2Responses();
+
+  /// actually perform writing
+  void doWrite();
+
+  Stream* createStream(int32_t sid, std::unique_ptr<HttpRequest>);
+  Stream* findStream(int32_t sid) const;
+
+  // may be used to signal a write from sendResponse
+  void signalWrite();
+
+ private:
+  static constexpr size_t kOutBufferLen = 64 * 1024 * 1024;
+  std::array<uint8_t, kOutBufferLen> _outbuffer;
```

Comment:
Does this mean we have a fixed size buffer of 64MB per H2CommTask? That seems a bit much. Can we allocate this buffer on the heap if needed?